File: src/lib/nodes/synthesis.ts
import { ChatGroq } from "@langchain/groq";
import { AgentState, DimensionScores, SynthesisRationales } from "../types";
import { safeParseLlmJson, extractTextContent } from "../utils/parseJson";

export const synthesisNode = async (state: AgentState): Promise<Partial<AgentState>> => {
  const companyInfo = state.companyInfo;

  if (!companyInfo?.name) {
    return {
      streamLog: ["⊗ ABORT — No entity resolved for Synthesis."],
      scores: getEmptyScores(),
      synthesis: getEmptySynthesis(),
    };
  }

  const resolvedName = companyInfo.name;

  const webSentimentScore = state.webAnalysis?.sentimentScore ?? 5;
  const financialHealthScore = state.financialAnalysis?.financialHealthScore ?? 5;
  const competitiveScore = state.competitiveAnalysis?.competitiveScore ?? 5;
  const redFlags = state.webAnalysis?.redFlags?.join(", ") || "None found";
  const moatType = state.competitiveAnalysis?.moatType || "none";
  const moatStrength = state.competitiveAnalysis?.moatStrength ?? 5;
  const valuationRisk = state.financialAnalysis?.valuationRisk || "unknown";
  const keyDevelopments = state.webAnalysis?.keyDevelopments?.join(", ") || "None found";

  try {
    const model = new ChatGroq({
      model: "llama-3.3-70b-versatile",
      temperature: 0.2,
      maxTokens: 2048,
    });

    const systemPrompt =
      "You are a partner at a top-tier investment firm.\nReturn ONLY valid JSON, no markdown.";
    const userPrompt = `Score ${resolvedName} across 5 investment dimensions.

Research data:
- Web sentiment score: ${webSentimentScore}/10
- Financial health score: ${financialHealthScore}/10
- Competitive score: ${competitiveScore}/10
- Red flags: ${redFlags}
- Moat: ${moatType}, strength ${moatStrength}/10
- Valuation risk: ${valuationRisk}
- Key developments: ${keyDevelopments}

Score each dimension 1-10 strictly based on data above.
Then calculate weightedTotal using:
Growth x 0.25 + Moat x 0.20 + FinancialHealth x 0.25 + Sentiment x 0.15 + Valuation x 0.15

Return this exact JSON:
{
  "scores": {
    "growth": number,
    "moat": number,
    "financialHealth": number,
    "sentiment": number,
    "valuation": number,
    "weightedTotal": number
  },
  "growthRationale": "string",
  "moatRationale": "string",
  "financialHealthRationale": "string",
  "sentimentRationale": "string",
  "valuationRationale": "string",
  "keyStrengths": ["string"],
  "keyRisks": ["string"]
}`;

    const response = await model.invoke([
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ]);

    const raw = extractTextContent(response.content);
    const { data: parsedResult, error } = safeParseLlmJson(raw, "SynthesisNode");

    const result = (parsedResult ?? {}) as Record<string, unknown>;
    if (error) {
      console.warn("[SynthesisNode] Using fallback scores due to parse error.");
    }

    const rawScores = (result.scores as Record<string, unknown>) || {};
    const s_growth = typeof rawScores.growth === "number" ? rawScores.growth : 5;
    const s_moat = typeof rawScores.moat === "number" ? rawScores.moat : 5;
    const s_fin = typeof rawScores.financialHealth === "number" ? rawScores.financialHealth : 5;
    const s_sent = typeof rawScores.sentiment === "number" ? rawScores.sentiment : 5;
    const s_val = typeof rawScores.valuation === "number" ? rawScores.valuation : 5;

    const calculated =
      s_growth * 0.25 + s_moat * 0.2 + s_fin * 0.25 + s_sent * 0.15 + s_val * 0.15;

    const scores: DimensionScores = {
      growth: s_growth,
      moat: s_moat,
      financialHealth: s_fin,
      sentiment: s_sent,
      valuation: s_val,
      weightedTotal: Number(calculated.toFixed(2)),
    };

    const synthesis: SynthesisRationales = {
      growthRationale: (result.growthRationale as string) || "No rationale provided",
      moatRationale: (result.moatRationale as string) || "No rationale provided",
      financialHealthRationale:
        (result.financialHealthRationale as string) || "No rationale provided",
      sentimentRationale: (result.sentimentRationale as string) || "No rationale provided",
      valuationRationale: (result.valuationRationale as string) || "No rationale provided",
      keyStrengths: Array.isArray(result.keyStrengths) ? result.keyStrengths as string[] : [],
      keyRisks: Array.isArray(result.keyRisks) ? result.keyRisks as string[] : [],
    };

    return {
      scores,
      synthesis,
      streamLog: [`◈ SCORE: ${scores.weightedTotal}/10`],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      streamLog: [`⊗ SYNTHESIS FAULT — ${errorMessage}`],
      scores: getEmptyScores(),
      synthesis: getEmptySynthesis(),
    };
  }
};

function getEmptyScores(): DimensionScores {
  return {
    growth: 5,
    moat: 5,
    financialHealth: 5,
    sentiment: 5,
    valuation: 5,
    weightedTotal: 5.0,
  };
}

function getEmptySynthesis(): SynthesisRationales {
  return {
    growthRationale: "Failed to evaluate",
    moatRationale: "Failed to evaluate",
    financialHealthRationale: "Failed to evaluate",
    sentimentRationale: "Failed to evaluate",
    valuationRationale: "Failed to evaluate",
    keyStrengths: [],
    keyRisks: [],
  };
}

File: backend/nodes/synthesis.py
"""
nodes/synthesis.py — Investment Synthesis & Scoring Node
Port of src/lib/nodes/synthesis.ts
"""

import logging
from langchain_groq import ChatGroq
from langchain_core.messages import SystemMessage, HumanMessage

from state import AgentState
from utils.parse_json import safe_parse_llm_json

logger = logging.getLogger(__name__)


def _empty_scores() -> dict:
    return {
        "growth": 5,
        "moat": 5,
        "financialHealth": 5,
        "sentiment": 5,
        "valuation": 5,
        "weightedTotal": 5.0,
    }


def _empty_synthesis() -> dict:
    return {
        "growthRationale": "Failed to evaluate",
        "moatRationale": "Failed to evaluate",
        "financialHealthRationale": "Failed to evaluate",
        "sentimentRationale": "Failed to evaluate",
        "valuationRationale": "Failed to evaluate",
        "keyStrengths": [],
        "keyRisks": [],
    }


def _safe_num(val: object, default: float = 5.0) -> float:
    return float(val) if isinstance(val, (int, float)) else default


async def synthesis_node(state: AgentState) -> dict:
    company_info = state.get("companyInfo") or {}

    if not company_info.get("name"):
        return {
            "streamLog": ["⊗ ABORT — No entity resolved for Synthesis."],
            "scores": _empty_scores(),
            "synthesis": _empty_synthesis(),
        }

    resolved_name = company_info["name"]

    web_analysis = state.get("webAnalysis") or {}
    financial_analysis = state.get("financialAnalysis") or {}
    competitive_analysis = state.get("competitiveAnalysis") or {}

    web_sentiment_score = _safe_num(web_analysis.get("sentimentScore"), 5)
    financial_health_score = _safe_num(financial_analysis.get("financialHealthScore"), 5)
    competitive_score = _safe_num(competitive_analysis.get("competitiveScore"), 5)
    red_flags = ", ".join(web_analysis.get("redFlags") or []) or "None found"
    moat_type = competitive_analysis.get("moatType") or "none"
    moat_strength = _safe_num(competitive_analysis.get("moatStrength"), 5)
    valuation_risk = financial_analysis.get("valuationRisk") or "unknown"
    key_developments = ", ".join(web_analysis.get("keyDevelopments") or []) or "None found"

    try:
        model = ChatGroq(
            model="llama-3.3-70b-versatile",
            temperature=0.2,
            max_tokens=2048,
        )

        system_prompt = (
            "You are a partner at a top-tier investment firm.\n"
            "Return ONLY valid JSON, no markdown."
        )
        user_prompt = f"""Score {resolved_name} across 5 investment dimensions.

Research data:
- Web sentiment score: {web_sentiment_score}/10
- Financial health score: {financial_health_score}/10
- Competitive score: {competitive_score}/10
- Red flags: {red_flags}
- Moat: {moat_type}, strength {moat_strength}/10
- Valuation risk: {valuation_risk}
- Key developments: {key_developments}

Score each dimension 1-10 strictly based on data above.
Then calculate weightedTotal using:
Growth x 0.30 + Moat x 0.25 + Sentiment x 0.20 + FinancialHealth x 0.15 + Valuation x 0.10

Return this exact JSON:
{{
  "scores": {{
    "growth": number,
    "moat": number,
    "financialHealth": number,
    "sentiment": number,
    "valuation": number,
    "weightedTotal": number
  }},
  "growthRationale": "string",
  "moatRationale": "string",
  "financialHealthRationale": "string",
  "sentimentRationale": "string",
  "valuationRationale": "string",
  "keyStrengths": ["string"],
  "keyRisks": ["string"]
}}"""

        response = await model.ainvoke([
            SystemMessage(content=system_prompt),
            HumanMessage(content=user_prompt),
        ])
        raw = response.content if isinstance(response.content, str) else str(response.content)
        parsed_result, error = safe_parse_llm_json(raw, "SynthesisNode")

        result = parsed_result or {}
        if error:
            logger.warning("[SynthesisNode] Using fallback scores due to parse error.")

        raw_scores = result.get("scores") or {}
        s_growth = _safe_num(raw_scores.get("growth"), 5)
        s_moat = _safe_num(raw_scores.get("moat"), 5)
        s_fin = _safe_num(raw_scores.get("financialHealth"), 5)
        s_sent = _safe_num(raw_scores.get("sentiment"), 5)
        s_val = _safe_num(raw_scores.get("valuation"), 5)

        # Recalculate locally to ensure accuracy
        calculated = round(s_growth * 0.30 + s_moat * 0.25 + s_sent * 0.20 + s_fin * 0.15 + s_val * 0.10, 2)

        scores = {
            "growth": s_growth,
            "moat": s_moat,
            "financialHealth": s_fin,
            "sentiment": s_sent,
            "valuation": s_val,
            "weightedTotal": calculated,
        }

        synthesis = {
            "growthRationale": result.get("growthRationale") or "No rationale provided",
            "moatRationale": result.get("moatRationale") or "No rationale provided",
            "financialHealthRationale": result.get("financialHealthRationale") or "No rationale provided",
            "sentimentRationale": result.get("sentimentRationale") or "No rationale provided",
            "valuationRationale": result.get("valuationRationale") or "No rationale provided",
            "keyStrengths": result.get("keyStrengths") if isinstance(result.get("keyStrengths"), list) else [],
            "keyRisks": result.get("keyRisks") if isinstance(result.get("keyRisks"), list) else [],
        }

        return {
            "scores": scores,
            "synthesis": synthesis,
            "streamLog": [f"◈ SCORE: {calculated}/10"],
        }

    except Exception as exc:
        error_message = str(exc)
        return {
            "streamLog": [f"⊗ SYNTHESIS FAULT — {error_message}"],
            "scores": _empty_scores(),
            "synthesis": _empty_synthesis(),
        }
